File: scala/atbash-cipher/src/main/scala/AtbashCipher.scala
object AtbashCipher {
  private val alphabet = 'a' to 'z'
  private val chunkSize = 5

  def encode(plainText: String): String =
    plainText
      .map(convert(_, alphabet))
      .collect { case Some(c) => c }
      .sliding(chunkSize, chunkSize)
      .map(_.mkString)
      .mkString(" ")

  def decode(cipherText: String): String =
    cipherText
      .replaceAll(" ", "")
      .map(convert(_, alphabet.reverse))
      .collect { case Some(c) => c }
      .mkString

  private def convert(char: Char, key: Seq[Char]) = char match {
    case c if c.isDigit => Some(c)
    case c if c.isLetter => Some(key.reverse(key.indexOf(c.toLower)))
    case _ => None
  }
}

File: fsharp/atbash-cipher/AtbashCipher.fs
module AtbashCipher

open System

let private convert key c =
    let convertLetter l =
        let index = List.findIndex ((=) l) key
        let lookup = List.rev key
        lookup.[index]

    match Char.ToLower(c) with
    | d when Char.IsDigit(d) -> Some d
    | l when Char.IsLetter(l) -> Some (convertLetter l)
    | _ -> None

let encode (str: string) =
    let key = ['a' .. 'z']
    str
    |> Seq.choose (convert key)
    |> Seq.chunkBySize 5
    |> Seq.map (String.Concat)
    |> String.concat " "

let decode (str: string) =
    let key = ['a' .. 'z'] |> List.rev
    str
    |> Seq.choose (convert key)
    |> String.Concat